### PR TITLE
perf: devirtualize lexicographic comparator on block binary-search hot path

### LINKED ISSUE #220: perf: branchless key comparison in data block seek
## Context

Data block binary search in `data_block/iter.rs:86-92` uses a 3-way branch (`Ordering::Less / Equal / Greater`) for key comparison. This is a predictable-branch-miss hot path during point lookups — the CPU branch predictor can't help because keys are essentially random relative to the search target.

## What

- Replace 3-way match with branchless comparison using `cmov`-friendly patterns
- Use the "lower bound" binary search style: `if key < target { lo = mid + 1 } else { hi = mid }` — single comparison, single conditional move
- Benchmark: measure seek latency (ns/op) on 4KB blocks with 64B keys before/after
- Verify correctness with existing data block tests

## Implementation sketch

```rust
// Before (3-way branch):
match key.cmp(target) {
    Ordering::Less => lo = mid + 1,
    Ordering::Equal => return Some(mid),
    Ordering::Greater => hi = mid,
}

// After (branchless lower bound + post-check):
let go_right = key < target;
lo = if go_right { mid + 1 } else { lo };
hi = if go_right { hi } else { mid };
// After loop: check if keys[lo] == target
```

## Blocked by

None.

## Estimate

1d

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Block-level binary search in `partition_point` probes restart heads via a predicate that calls `dyn UserComparator::compare` — a vtable dispatch on every `O(log restarts)` step. The `is_lexicographic()` opt-in already exists on the trait and is honored in `compare_prefixed_slice` (the prefix branch of `compare_key`, where the lex path avoids a prefix+suffix concatenation allocation), but the binary search itself routed every probe through vtable.

This PR branches once on `is_lexicographic()` at each predicate construction site and selects a closure that does direct slice comparison (`<[u8]>::cmp`) on the lex path. Each closure is a distinct type, so the generic `partition_point` / `seek` / `seek_upper` monomorphize per shape — the inner binary-search loop is virtual-call-free on the lex path. The custom-comparator branch is preserved verbatim.

## Sites devirtualized

| File | Functions | Predicates |
|---|---|---|
| `src/table/data_block/iter.rs` | `seek_to_key_seqno`, `seek_upper`, `seek_upper_exclusive` | 3 × 2 monomorphizations |
| `src/table/index_block/iter.rs` | `seek_with_cache_resets`, `seek_upper_impl` (covers both public `seek_upper` and `seek_upper_bound_cursor`) | 4 × 2 monomorphizations |

Each predicate construction site hoists ONE `is_lexicographic()` call out of the BS loop, amortising it across all `O(log restarts)` probes. The lex closure does direct `<[u8]>::cmp`; the dyn closure preserves the original `cmp.compare(...)` call.

## Out of scope (deliberate)

- **`partition_point` itself** already uses the canonical 2-way lower-bound shape (`if pred { left = mid+1 } else { right = mid }`) — LLVM emits cmov. No 3-way branch to remove.
- **3-way `Ordering` match in the linear-scan loops** inside restart intervals stays — `Equal`/`Greater` early-exit beats branchless scan-to-end-and-post-check for typical `restart_interval=8-16` scans.
- **`ParsedItem::compare_key` no-prefix branch** intentionally NOT devirtualized. An earlier revision added a lex fast path there too; review feedback correctly identified that the no-prefix branch has no allocation to avoid (the key is already contiguous), so an `is_lexicographic()` check before `compare()` would cost custom comparators a second vtable per linear-scan step without any matching saving on the default path (`<[u8]>::cmp` and `DefaultUserComparator::compare` both lower to memcmp). The lex fast path is kept only where it saves real work: `compare_prefixed_slice_lexicographic` (skips an allocation on the prefix branch) and the iter.rs BS predicate factories above.

## Test plan

- [x] `cargo nextest run` — **666 / 666 lib tests pass** + **10 devirt regression tests**
- [x] `cargo clippy --all-targets --lib --tests -- -D warnings` — clean
- [x] **Direct devirt verification** via counting-comparator wrapper that tracks `compare()` AND `is_lexicographic()` calls separately:
  - **Lex-path tests** (data_block × 3 entry points; index_block × 3 entry points incl. `seek_upper_bound_cursor`) — assert `count <= LEX_PATH_LINEAR_SCAN_BOUND` (≈ 2, only linear-scan compare_key contribution) AND `is_lex_count >= 1` (the BS predicate factory consulted is_lex). A regression where the dyn closure leaked into the BS predicate would produce `>= DYN_MIN_BS_PROBES` (= 7, = log₂(128)) calls — orders-of-magnitude separation from the bound.
  - **Dyn-path tests** (one per entry point) — assert `delta >= DYN_MIN_BS_PROBES` (= 7) on the BS-dominated block (128 keys × restart_interval=1 + above-max needle bounds linear-scan to 1 call). Distinguishes a working dyn BS (≥ 7 + 1) from a lex-leak (0 + 1).
  - **Equivalence tests** (data_block: seek + seek_upper + seek_upper_exclusive; index_block: seek + seek_upper + seek_upper_bound_cursor) across 6 boundary needles (empty slice = below-min, exact-min, between-keys 9-byte, exact-mid, exact-tail, above-max 9-byte) — proves lex and dyn paths produce identical landing positions for every `partition_point` boundary case (`left==0`, `left==len`, exact-hit), and surfaces a wrong-operator regression in any lex closure as a landing mismatch.
- [x] 41 existing `custom_comparator*` integration tests exercise the dyn path — all pass unchanged

## Benchmarks

`tools/db_bench --num 200000`, 5 runs per workload per branch (medians + ranges), measured on the initial devirtualization commit:

| Workload | main (med) | this PR (med) | Δ | main range | this PR range | Read |
|---|---:|---:|---:|---:|---:|---|
| fillseq | 1,717,606 | 1,675,953 | **-2.4%** | 1.57M-1.77M | 1.60M-1.73M | noise (write path, not on devirt) |
| readseq | 2,497,713 | 2,546,882 | **+2.0%** | 2.34M-2.90M | 2.12M-2.59M | noise (readseq does not binary-search per op) |
| readrandom | 370,160 | 373,223 | **+0.8%** | 362k-392k | 355k-380k | tiny win, inside noise |
| seekrandom | 276,825 | 284,261 | **+2.7%** | 271k-287k | 276k-303k | borderline positive signal |
| prefixscan | 160,785 | 148,139 | **-7.9%** | 154k-164k | 147k-172k | wide range — one this-PR run at 172k > all main runs |

**Honest read**: on a 200k-op macOS bench, expected win from saving `N-1` vtable calls per seek (`N = log2(restart_heads) ≈ 2-3`) is **0.5-2%** in absolute best case — below the noise floor of single-run db_bench on shared hardware. The signal is consistent with the theoretical prediction.

**The devirtualization itself is verified directly by test, not inferred from bench numbers** — the lex-path / dyn-path counting assertions (above) deterministically prove the right closure runs at each entry point. Real workloads with hot point-read paths and warm caches should see a larger relative benefit than this small synthetic bench.

Closes #220

## Summary by CodeRabbit

* **Performance**
  * Added lexicographic fast paths to iterator seek operations in table data and index blocks, reducing runtime overhead during binary-search probes.

* **Tests**
  * Added a suite of regression and devirtualization tests that count comparator usage and validate seek behavior across multiple boundary cases.

* **Documentation**
  * Clarified comments around comparator fast-path responsibilities to explain where lexicographic optimizations are applied.

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/coordinode-lsm-tree/pull/266?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/table/data_block/iter_test.rs [hunk 1]
```diff
@@ -1420,4 +1420,290 @@ mod tests {
...
+            fn name(&self) -> &'static str {
+                "counting"
+            }
+            fn compare(&self, a: &[u8], b: &[u8]) -> std::cmp::Ordering {
+                self.count.fetch_add(1, AtomicOrdering::Relaxed);
+                a.cmp(b)
+            }
+            fn is_lexicographic(&self) -> bool {
+                self.lex
+            }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/data_block/iter_test.rs:1494]: @coderabbitai Fixed in 395df958 — added `is_lex_count` to `CountingComparator` (counts `is_lexicographic()` vtable invocations separately) and added per-entry-point assertions that the lex branch is actually consulted (`is_lex_count >= 1`). This rules out the previously-uncovered regression where the lex branch was never selected and `compare_count == 0` passed for the wrong reason.

File: src/table/data_block/iter_test.rs [hunk 2]
```diff
@@ -1420,4 +1420,406 @@ mod tests {
...
+                    lex_seek, dyn_seek,
+                    "seek result must match for needle {label} ({needle:?})",
+                );
+                assert_eq!(
+                    lex_landing, dyn_landing,
+                    "landing position must match for needle {label} ({needle:?})",
+                );
+            }
+            Ok(())
+        }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/data_block/iter_test.rs:1899]: @coderabbitai Fixed in b33a34d5 — the data_block equivalence loop now runs the same six-needle table through `seek`, `seek_upper`, AND `seek_upper_exclusive`, comparing both the bool result and the landed user_key for each. A wrong operator in any lex closure would now surface as a landing mismatch.

File: src/table/index_block/mod.rs [hunk 3]
```diff
@@ -291,4 +303,308 @@ mod tests {
...
+                let dyn_landing = dyn_iter
+                    .next()
+                    .map(|h| h.materialize(index_block.as_slice()).end_key().clone());
+                assert_eq!(
+                    lex_landing.as_ref().map(|s| s.as_ref().to_vec()),
+                    dyn_landing.as_ref().map(|s| s.as_ref().to_vec()),
+                    "index landing end_key must match for needle {label} ({needle:?})",
+                );
+            }
+        }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/index_block/mod.rs:746]: @coderabbitai Fixed in b33a34d5 — the index_block equivalence loop now also runs `seek_upper` (exercises `seek_upper_impl`'s strict-`<` branch at restart_interval == 1 with check_back_cache, distinct from `seek`'s 3-way seqno-aware predicate). Boundary needles + landed end_key bytes compared between lex and dyn.

File: src/table/index_block/mod.rs [hunk 4]
```diff
@@ -291,4 +301,437 @@ mod tests {
...
+                    .map(|h| h.materialize(index_block.as_slice()).end_key().clone());
+                assert_eq!(
+                    lex_landing.as_ref().map(|s| s.as_ref().to_vec()),
+                    dyn_landing.as_ref().map(|s| s.as_ref().to_vec()),
+                    "index seek landing must match for needle {label} ({needle:?})",
+                );
+
+                // seek_upper (reverse upper-bound — exercises seek_upper_impl
+                // with check_back_cache=true, strict-`<` predicate at
+                // restart_interval == 1).
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/table/index_block/mod.rs]: @coderabbitai Fixed in 267f6f40 — reworded both seek_upper / seek_upper_bound_cursor comments to reflect that this test block uses `restart_interval = 4`, so both variants land in the shared `restart_interval > 1` arm (same `<=` predicate). The divergent `restart_interval == 1` predicates are covered separately by the BS-dominated tests above which build the block with `restart_interval = 1`.
